src/Feldspar/Onnx/onnxToFeld.hs

```diff
@@ -29,67 +29,111 @@
+import qualified Data.Foldable as D (toList, foldMap)
+import qualified Data.Set as S
 import Data.List (intercalate)
-import Data.Maybe (maybe, fromJust, fromMaybe)
+import Data.Maybe (fromJust, fromMaybe)
 import System.Environment (getArgs)
+import System.FilePath (takeBaseName, (<.>))
+import System.IO (IOMode(WriteMode), BufferMode(BlockBuffering), openFile, hClose, hPutStr
+                 , hSetBuffering, hSetBinaryMode)
 
 main :: IO ()
 main = do args <- getArgs
           let [modelFileName] = take 1 args -- First argument is file name
-          -- putStrLn modelFileName
+              modelBaseName = takeBaseName modelFileName
+              dataFileName = modelBaseName <.> "data"
+              progFileName = modelBaseName <.> "hs"
           modelBinary <- L.readFile modelFileName
-          -- putStrLn "Read it!"
           let model = either (\s -> error $ "Error: " ++ s) fst $ messageGet modelBinary
-              gr = fromMaybe (error "No graph in model") $ graph model
-          -- putStrLn $ maybe (error "Graph has no name") toStr $ G.name gr
-          let graphNodes = D.toList $ node gr
-          -- putStrLn $ show $ length graphNodes
-          -- putStrLn "module Inits where"
-          -- putStrLn "import Feldspar"
-          -- putStr   "-- Num graph initializers "
-          -- putStrLn $ show $ length $ D.toList $ G.initializer gr
-          -- putStrLn "-- Graph initializers"
-          -- putStrLn "typedef float Float;"
-          -- putStrLn $ unlines $ map showGIc $ take 1000 $ D.toList $ G.initializer gr
-          putStrLn "Graph inputs"
-          putStrLn $ unlines $ map showVI $ D.toList $ G.input gr
-          putStrLn "Graph nodes"
-          putStr $ unlines $ map showNode graphNodes
-
-showNode :: NodeProto -> String
-showNode n = "  " ++ outs ++ " = " ++ op ++ " " ++ ins ++ "\n" ++ attrs
-  where ins = unwords $ map toStr $ D.toList $ N.input n
-        -- nameStr = maybe (error "No name string") toStr $ N.name n
-        op = "onnx" ++ toStr (fromJust $ N.op_type n)
-        outs = unwords $ map toStr $ D.toList $ N.output n
-        attrs = unlines $ map showAttribute $ D.toList $ N.attribute n
-
-showAttribute :: AttributeProto -> String
-showAttribute a = "  -- " ++ nStr ++ " " ++ val nStr
+              gr = fromMaybe (error "No graph in model") $ O.graph model
+
+          -- Write the weights
+          dfile <- openFile dataFileName WriteMode
+          hSetBinaryMode dfile True
+          hSetBuffering dfile $ BlockBuffering Nothing
+          B.hPutBuilder dfile $ D.foldMap buildInitTensor $ G.initializer gr
+          hClose dfile
+
+          -- Write the program
+          pfile <- openFile progFileName WriteMode
+          hPutStr pfile $ mkProgramFile gr
+          hClose pfile
+
+-- | Extract initialized tensors
+buildInitTensor :: TP.TensorProto -> B.Builder
+buildInitTensor t = B.stringUtf8 (unwords $ showElemT dt : show (length ds) : map show ds)
```

Comment:
Does the content of the characters in the data file fit in Char8? Large parts of the Prelude is available in `Data.ByteString.Char8` for ByteString, `unwords` is one of them.

Reply:
Yes, but the return type of 'B.stringUtf8' is not a ByteString, it is a ByteString.Builder, and that type does not support e.g. unwords. Using builders allow for zero (user space) copy output.

Reply:
Ah, yes, changing the type of the stuff in the parenthesis requires switching `B.stringUtf8` to `B.byteString`.